src/plugins/hiddenColumns/contextMenuItem/showColumn.js

```diff
@@ -6,112 +6,100 @@ import * as C from '../../../i18n/constants';
+        startVisualColumn === endVisualColumn;
+
+      let startPhysicalColumn = this.toPhysicalColumn(startVisualColumn);
+      let endPhysicalColumn = this.toPhysicalColumn(endVisualColumn);
+
+      if (onlyFirstVisibleColumnSelected) {
+        startPhysicalColumn = 0;
+      }
 
-      } else {
-        rangeEach(start, end, column => hiddenColumnsPlugin.showColumn(column));
+      if (onlyLastVisibleColumnSelected) {
+        endPhysicalColumn = this.countSourceCols() - 1;
       }
 
+      hiddenColumnsPlugin.showColumns(columns);
+
+      const startVisualColumnAfterAction = this.toVisualColumn(startPhysicalColumn);
+      const endVisualColumnAfterAction = this.toVisualColumn(endPhysicalColumn);
+
+      // Selection start and selection end coordinates might be changed after showing some items.
+      this.selectColumns(startVisualColumnAfterAction, endVisualColumnAfterAction);
       this.render();
       this.view.wt.wtOverlays.adjustElementsSize(true);
     },
     disabled: false,
     hidden() {
-      if (!hiddenColumnsPlugin.hiddenColumns.length || !this.selection.isSelectedByColumnHeader()) {
+      if (!this.selection.isSelectedByColumnHeader() || hiddenColumnsPlugin.getHiddenColumns().length < 1) {
         return true;
       }
 
-      beforeHiddenColumns.length = 0;
-      afterHiddenColumns.length = 0;
+      columns.length = 0;
 
-      const { from, to } = this.getSelectedRangeLast();
-      const start = Math.min(from.col, to.col);
-      const end = Math.max(from.col, to.col);
-      let hiddenInSelection = false;
+      const [, startColumn, , endColumn] = this.getSelectedLast();
+      const start = Math.min(startColumn, endColumn);
+      const end = Math.max(startColumn, endColumn);
+      const sequence = this.columnIndexMapper.getIndexesSequence();
+      const physicalStartColumn = this.columnIndexMapper.getPhysicalFromVisualIndex(start);
+      const currentStartPosition = sequence.indexOf(physicalStartColumn);
 
       if (start === end) {
-        let totalColumnLength = this.countSourceCols();
+        const columnsBefore = [];
+        const columnsAfter = [];
 
-        rangeEach(0, totalColumnLength, (column) => {
-          const partedHiddenLength = beforeHiddenColumns.length + afterHiddenColumns.length;
+        rangeEach(0, currentStartPosition - 1, (column) => {
+          if (hiddenColumnsPlugin.isHidden(sequence[column], true)) {
+            columnsBefore.push(sequence[column]);
+
+          } else {
+            columnsBefore.length = 0;
 
-          if (partedHiddenLength === hiddenColumnsPlugin.hiddenColumns.length) {
             return false;
           }
+        });
 
-          if (column < start && hiddenColumnsPlugin.isHidden(column)) {
-            beforeHiddenColumns.push(column);
+        rangeEach(currentStartPosition + 1, this.countCols() - 1, (column) => {
+          if (hiddenColumnsPlugin.isHidden(sequence[column], true)) {
```

Comment:
`isHidden` is used with the second parameter, this is obsolete and not used in hiddenColumns plugin. We should double-check if this works as expected.
```suggestion
          if (hiddenColumnsPlugin.isHidden(sequence[column])) {
```

Reply:
As above.